src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs

```diff
@@ -133,34 +133,97 @@ public ItemData(I item, int elementOrder, bool conditionResult)
             public bool ConditionResult { get { return _conditionResult; } }
         }
 
+        private class OperationWithCache : IItemOperation
+        {
+            private readonly IItemOperation _operation;
+            private readonly Lazy<Dictionary<ISet<string>, ImmutableList<ItemData>>> _cache;
+
+            public OperationWithCache(IItemOperation operation)
+            {
+                _operation = operation;
+                _cache = new Lazy<Dictionary<ISet<string>, ImmutableList<ItemData>>>(
+                    () => new Dictionary<ISet<string>, ImmutableList<ItemData>>());
+            }
+
+            public void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
+            {
+                _operation.Apply(listBuilder, globsToIgnore);
+            }
+
+            public bool TryGetFromCache(ISet<string> globsToIgnore, out ImmutableList<ItemData> items)
+            {
+                return _cache.Value.TryGetValue(globsToIgnore, out items);
+            }
+
+            public void AddItemsToCache(ImmutableHashSet<string> globsToIgnore, ImmutableList<ItemData> items)
+            {
+                _cache.Value[globsToIgnore] = items;
+            }
+        }
 
         private class LazyItemList
         {
             private readonly LazyItemList _previous;
-            private readonly LazyItemOperation _operation;
+            private readonly OperationWithCache _operation;
 
             public LazyItemList(LazyItemList previous, LazyItemOperation operation)
             {
                 _previous = previous;
-                _operation = operation;
+                _operation = new OperationWithCache(operation);
             }
 
             public ImmutableList<ItemData>.Builder GetItems(ImmutableHashSet<string> globsToIgnore)
             {
-                //  TODO: Check for cached results
+                // Cache results only on the operation whose results are required by an external caller (via GetItems)
+                // If the head of this LazyItemList is uncached, then the tail either:
+                //   - has all operations uncached (nobody needed their results until now). In this case we have to compute the head plus the entire tail,
+                // but we don't need to cache each node of the tail because no external caller wanted their results
+                //   - has some operations cached (somebody needed their result at some point). In this case we have to compute the head plus the part of the 
+                // tail up to the first cached operation.
```

Comment:
Yes, that would be the worst case. It happens if you have this:

```xml
1: <C Include="a"/>
2: <ForcesCacheOnC Include="@(C)"/>
3: <C Include="b"/>
4: <ForcesCacheOnC Include="@(C)"/>
5: <C Include="c"/>
6: <ForcesCacheOnC Include="@(C)"/>
```

However, the caches will share structural state. So the cache at operation `3` will share common state with the previously made cache at operation `1` (shared state is the collection with item 'a'). This is done via the immutable list. Also, when `GetItems` is called on operation `3`, it won't redo operation `1` but use its cache.

So no computation gets ever redone, but we will have a lot of `OperationWithCache` decorators if the above pattern is very common.

We could do other fancy things like reference counting (assuming an external caller calls `GetItems` only once, then we could decrement some CalleeNumber on each call per operation, and then `null` the cache for that operation when it has satisfied all of it's external callers). But I don't think it's worth making it more complex at this point until we get an outside ask about it :)